macos/Sources/Features/Terminal/BaseTerminalController.swift

```diff
@@ -364,11 +369,61 @@ class BaseTerminalController: NSWindowController,
 
     // MARK: Split Tree Management
 
+    private func containsInSurfaceTree(
+        _ surface: Ghostty.SurfaceView?,
+        tree: SplitTree<Ghostty.SurfaceView>? = nil
+    ) -> Bool {
+        guard let surface else { return false }
+        let activeTree = tree ?? surfaceTree
+        return activeTree.contains(surface)
+    }
+
+    private func pruneRecentFocusedSurface(tree: SplitTree<Ghostty.SurfaceView>? = nil) {
+        guard recentFocusedSurface != nil else { return }
+
+        if !containsInSurfaceTree(recentFocusedSurface, tree: tree) {
+            recentFocusedSurface = nil
+        }
+    }
+
+    private func updateRecentFocusedSurface(
+        previous: Ghostty.SurfaceView?,
+        current: Ghostty.SurfaceView?
+    ) {
+        guard let previous, let current else { return }
+        guard previous !== current else { return }
+        guard containsInSurfaceTree(previous), containsInSurfaceTree(current) else { return }
+
+        recentFocusedSurface = previous
+    }
+
+    func splitFocusTarget(
+        for direction: Ghostty.SplitFocusDirection,
+        from target: Ghostty.SurfaceView
+    ) -> Ghostty.SurfaceView? {
+        // Find the node for the target surface
+        guard let targetNode = surfaceTree.root?.node(view: target) else { return nil }
+
+        switch direction {
+        case .recent:
+            pruneRecentFocusedSurface()
+
+            guard let recentSurface = recentFocusedSurface, recentSurface !== target else {
+                return nil
+            }
+
+            return containsInSurfaceTree(recentSurface) ? recentSurface : nil
+
+        default:
+            return surfaceTree.focusTarget(for: direction.toSplitTreeFocusDirection(), from: targetNode)
+        }
+    }
+
     /// Find the next surface to focus when a node is being closed.
     /// Goes to previous split unless we're the leftmost leaf, then goes to next.
     private func findNextFocusTargetAfterClosing(node: SplitTree<Ghostty.SurfaceView>.Node) -> Ghostty.SurfaceView? {
         guard let root = surfaceTree.root else { return nil }
-        
+
```

Comment:
My editor trims trailing spaces on save...  Let me know if these kind of changes should be avoided.